### PR TITLE
enable user to select kdf function

### PR DESCRIPTION
**Pull request by jenpaff**
by default we use pbkdf2 for enhanced security

### CODE REVIEW

File: src/cli/existing_mnemonic.rs [hunk 1]
```diff
@@ -79,6 +81,16 @@ pub fn subcommand<'a, 'b>() -> App<'a, 'b> {
                 and ^(0x00[a-fA-F0-9]{62})$ for BLS withdrawal credentials.",
                 ),
         )
+        .arg(
+            Arg::with_name("kdf")
+                .long("kdf")
+                .required(false)
+                .takes_value(true)
+                .help(
```
Comment (mksh) [inline src/cli/existing_mnemonic.rs:89]: I believe it's also useful to specify possible values here via `.possible_values()` as described in https://github.com/clap-rs/clap/blob/v2.34.0/examples/11_only_specific_values.rs#L21

File: src/key_material.rs [hunk 2]
```diff
@@ -225,4 +236,101 @@ mod test {
...
+        assert_eq!(
+            "5b17290ac4e36c497378351ed61253a239d408466918fe5a8dd79b28ef67f9df",
+            hex::encode(secret_key.as_bytes())
+        );
+        let key_material_3 = keystores.get(2).unwrap();
+        assert_eq!(key_material_3.clone().keystore.unwrap().pubkey(), "8bf0a669a51d0cb6ff745e4b0aa7c41e8de8d179ff9267977e76c7188aaa2fb1b8b1bdfefcc545d9efdac0b4bc2239e6");
+        let secret_key = &key_material_3.voting_secret;
+        assert_eq!(
+            "1cb46382c3c046f7817298d1ad3454a7175be6e53059bdd287e8fcbf4c6fb2e8",
+            hex::encode(secret_key.as_bytes())
```
Comment (mksh) [inline src/key_material.rs:333]: This should also test that generated keystore has proper encryption specified, e.g.

given generated keystore record is like
```
{
  "crypto": {
    "checksum": {
      "function": "sha256",
      "message": "b95e2bf5c5aafe0e9149f08d2e38f1daa5d20b73972433cd1b523e647249544c"
    },
    "cipher": {
      "function": "aes-128-ctr",
      "message": "bbed72f35ea11f66a44ca42c5b593574db1fb1bd763c91e64a8629899fb58ffe",
      "params": {
        "iv": "793b3201f800b017e80c35b5486ab5df"
      }
    },
    "kdf": {
      "function": "pbkdf2",
      "message": "",
      "params": {
        "dklen": 32,
        "n": 262144,
        "p": 1,
        "r": 8,
        "salt": "14260c0c72c4afcde2b99ccdf2e1b1257d3cb5dfdc19ddbdc70f2b36eee90b9a"
      }
    }
  },
  "description": "",
  "path": "m/12381/3600/9/0/0",
  "pubkey": "b9d798bbff9a7d7599617035c1141ef607f0b4398ad9c3753219c634397b0a6f6f162b280f9322e41e4878d3d1a351df",
  "uuid": "8e32aff9-96f2-4e87-80c7-0db7ab1f8327",
  "version": 4
}
```

the test logic should parse json and assert `/kdf/function` property is `"pbkdf2"`

File: src/key_material.rs [hunk 3]
```diff
@@ -225,4 +236,101 @@ mod test {
...
+        );
+
+        let key_material_3 = keystores.get(2).unwrap();
+        assert_eq!(key_material_3.clone().keystore.unwrap().pubkey(), "8bf0a669a51d0cb6ff745e4b0aa7c41e8de8d179ff9267977e76c7188aaa2fb1b8b1bdfefcc545d9efdac0b4bc2239e6");
+        let secret_key = &key_material_3.voting_secret;
+        assert_eq!(
+            "1cb46382c3c046f7817298d1ad3454a7175be6e53059bdd287e8fcbf4c6fb2e8",
+            hex::encode(secret_key.as_bytes())
+        );
```
Comment (mksh) [inline src/key_material.rs:286]: This should also test that generated keystore has proper encryption specified, e.g.

given generated keystore record is like
```
{
  "crypto": {
    "checksum": {
      "function": "sha256",
      "message": "b95e2bf5c5aafe0e9149f08d2e38f1daa5d20b73972433cd1b523e647249544c"
    },
    "cipher": {
      "function": "aes-128-ctr",
      "message": "bbed72f35ea11f66a44ca42c5b593574db1fb1bd763c91e64a8629899fb58ffe",
      "params": {
        "iv": "793b3201f800b017e80c35b5486ab5df"
      }
    },
    "kdf": {
      "function": "scrypt",
      "message": "",
      "params": {
        "dklen": 32,
        "n": 262144,
        "p": 1,
        "r": 8,
        "salt": "14260c0c72c4afcde2b99ccdf2e1b1257d3cb5dfdc19ddbdc70f2b36eee90b9a"
      }
    }
  },
  "description": "",
  "path": "m/12381/3600/9/0/0",
  "pubkey": "b9d798bbff9a7d7599617035c1141ef607f0b4398ad9c3753219c634397b0a6f6f162b280f9322e41e4878d3d1a351df",
  "uuid": "8e32aff9-96f2-4e87-80c7-0db7ab1f8327",
  "version": 4
}
```

the test logic should parse json and assert /kdf/function property is "scrypt"

File: tests/e2e/new_mnemonic.rs [hunk 4]
```diff
@@ -255,8 +372,8 @@ fn test_withdrawal_credentials_execution() -> Result<(), Box<dyn std::error::Err
     cmd.arg(decryption_password);
     cmd.arg("--num_validators");
     cmd.arg(num_validators);
-    cmd.arg("--withdrawal_credentials");
-    cmd.arg(execution_withdrawal_credentials);
+    cmd.arg("--kdf");
+    cmd.arg(kdf);
```
Comment (mksh) [inline tests/e2e/new_mnemonic.rs:376]: This should also test that generated keystore has proper encryption specified, e.g.

given generated keystore record is like
```
{
  "crypto": {
    "checksum": {
      "function": "sha256",
      "message": "b95e2bf5c5aafe0e9149f08d2e38f1daa5d20b73972433cd1b523e647249544c"
    },
    "cipher": {
      "function": "aes-128-ctr",
      "message": "bbed72f35ea11f66a44ca42c5b593574db1fb1bd763c91e64a8629899fb58ffe",
      "params": {
        "iv": "793b3201f800b017e80c35b5486ab5df"
      }
    },
    "kdf": {
      "function": "scrypt",
      "message": "",
      "params": {
        "dklen": 32,
        "n": 262144,
        "p": 1,
        "r": 8,
        "salt": "14260c0c72c4afcde2b99ccdf2e1b1257d3cb5dfdc19ddbdc70f2b36eee90b9a"
      }
    }
  },
  "description": "",
  "path": "m/12381/3600/9/0/0",
  "pubkey": "b9d798bbff9a7d7599617035c1141ef607f0b4398ad9c3753219c634397b0a6f6f162b280f9322e41e4878d3d1a351df",
  "uuid": "8e32aff9-96f2-4e87-80c7-0db7ab1f8327",
  "version": 4
}
```

the test logic should parse json and assert /kdf/function property is "scrypt"

File: tests/e2e/new_mnemonic.rs [hunk 5]
```diff
@@ -198,7 +259,64 @@ fn test_withdrawal_credentials_bls() -> Result<(), Box<dyn std::error::Error>> {
...
+    cmd.arg("--chain");
+    cmd.arg(chain);
+    cmd.arg("--keystore_password");
+    cmd.arg(decryption_password);
+    cmd.arg("--num_validators");
+    cmd.arg(num_validators);
+    cmd.arg("--kdf");
+    cmd.arg(kdf);
 
     cmd.assert().success();
```
Comment (mksh) [inline tests/e2e/new_mnemonic.rs:321]: This should also test that generated keystore has proper encryption specified, e.g.

given generated keystore record is like
```
{
  "crypto": {
    "checksum": {
      "function": "sha256",
      "message": "b95e2bf5c5aafe0e9149f08d2e38f1daa5d20b73972433cd1b523e647249544c"
    },
    "cipher": {
      "function": "aes-128-ctr",
      "message": "bbed72f35ea11f66a44ca42c5b593574db1fb1bd763c91e64a8629899fb58ffe",
      "params": {
        "iv": "793b3201f800b017e80c35b5486ab5df"
      }
    },
    "kdf": {
      "function": "pbkdf2",
      "message": "",
      "params": {
        "dklen": 32,
        "n": 262144,
        "p": 1,
        "r": 8,
        "salt": "14260c0c72c4afcde2b99ccdf2e1b1257d3cb5dfdc19ddbdc70f2b36eee90b9a"
      }
    }
  },
  "description": "",
  "path": "m/12381/3600/9/0/0",
  "pubkey": "b9d798bbff9a7d7599617035c1141ef607f0b4398ad9c3753219c634397b0a6f6f162b280f9322e41e4878d3d1a351df",
  "uuid": "8e32aff9-96f2-4e87-80c7-0db7ab1f8327",
  "version": 4
}
```

the test logic should parse json and assert /kdf/function property is "pbkdf2"

**Review by mksh — CHANGES_REQUESTED** [review]
Mostly good, some additional testing logic (to test keystore generated indeed uses kdf algorithm equals to requested)

File: src/cli/existing_mnemonic.rs (hunk 1 above)
Comment (jenpaff) [reply src/cli/existing_mnemonic.rs:91]: added in [9ea991f](https://github.com/ChorusOne/eth-staking-smith/pull/18/commits/9ea991ffa7318b5027c07b6d17421f0e83be3f23)

File: src/cli/new_mnemonic.rs [hunk 6]
```diff
@@ -46,13 +46,22 @@ pub fn subcommand<'a, 'b>() -> App<'a, 'b> {
...
+                and ^(0x00[a-fA-F0-9]{62})$ for BLS withdrawal credentials.",
+                ),
+        )
+        .arg(
+            Arg::with_name("kdf")
+                .long("kdf")
+                .required(false)
+                .takes_value(true)
+                .help(
+                    "Use this argument to select the key derivation function for the keystores depending on your use case with `scrypt` using higher security parameters
```
[the remark this answers is not in the document]
Comment (jenpaff) [reply src/cli/new_mnemonic.rs:65]: added in [9ea991f](https://github.com/ChorusOne/eth-staking-smith/pull/18/commits/9ea991ffa7318b5027c07b6d17421f0e83be3f23)

File: src/key_material.rs (hunk 2 above)
Comment (jenpaff) [reply src/key_material.rs]: added in [2064c11](https://github.com/ChorusOne/eth-staking-smith/pull/18/commits/2064c11669605549e442830e780a2dff7ff6f50a)

File: src/key_material.rs (hunk 3 above)
Comment (jenpaff) [reply src/key_material.rs]: added in [2064c11](https://github.com/ChorusOne/eth-staking-smith/pull/18/commits/2064c11669605549e442830e780a2dff7ff6f50a)

File: tests/e2e/new_mnemonic.rs (hunk 5 above)
Comment (jenpaff) [reply tests/e2e/new_mnemonic.rs:325]: added in [d939b31](https://github.com/ChorusOne/eth-staking-smith/pull/18/commits/d939b31a75957c2f2769a54fdd8ff69cadb9579d)

**Review by mksh — APPROVED** [review]
LGTM

**Pull request merged**